torch/autograd/gradcheck.py

```diff
@@ -368,81 +541,35 @@ def fail_test(msg):
         return False
 
     tupled_inputs = _as_tuple(inputs)
-    if not check_sparse_nnz and any(t.is_sparse for t in tupled_inputs if isinstance(t, torch.Tensor)):
-        return fail_test('gradcheck expects all tensor inputs are dense when check_sparse_nnz is set to False.')
 
-    # Make sure that gradients are saved for at least one input
-    any_input_requiring_grad = False
-    for idx, inp in enumerate(tupled_inputs):
-        if is_tensor_like(inp) and inp.requires_grad:
-            if not (inp.dtype == torch.float64 or inp.dtype == torch.complex128):
-                warnings.warn(
-                    f'Input #{idx} requires gradient and '
-                    'is not a double precision floating point or complex. '
-                    'This check will likely fail if all the inputs are '
-                    'not of double precision floating point or complex. ')
-            content = inp._values() if inp.is_sparse else inp
-            # TODO: To cover more problematic cases, replace stride = 0 check with
-            # "any overlap in memory" once we have a proper function to check it.
-            if content.layout is not torch._mkldnn:  # type: ignore
-                if not all(st > 0 or sz <= 1 for st, sz in zip(content.stride(), content.size())):
-                    raise RuntimeError(
-                        'The {}th input has a dimension with stride 0. gradcheck only '
-                        'supports inputs that are non-overlapping to be able to '
-                        'compute the numerical gradients correctly. You should call '
-                        '.contiguous on the input before passing it to gradcheck.')
-            any_input_requiring_grad = True
-            inp.retain_grad()
-    if not any_input_requiring_grad:
-        raise ValueError(
-            'gradcheck expects at least one input tensor to require gradient, '
-            'but none of the them have requires_grad=True.')
+    if not check_inputs(fail_test, tupled_inputs, check_sparse_nnz):
+        return False
 
     func_out = func(*tupled_inputs)
-    output = _differentiable_outputs(func_out)
-
-    if not output:
-        for i, o in enumerate(func_out):
-            def fn(input):
-                return _as_tuple(func(*input))[i]
-            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
-            for n in numerical:
-                if torch.ne(n, 0).sum() > 0:
-                    return fail_test('Numerical gradient for function expected to be zero')
-        return True
+    outputs = _differentiable_outputs(func_out)
 
-    for i, o in enumerate(output):
-        if not o.requires_grad:
-            continue
+    check_outputs(outputs)
```

Comment:
Do we have tests for gradcheck? I noticed some logic being moved around (like `check_outputs`).

Reply:
Maybe there are some scattered around - definitely not enough though. For example `raise_exception=False` isn't tested at all.